### PR TITLE
feat: ACI-1199 multithreaded download

### PR DESCRIPTION
**Pull request by duplabe**

### CODE REVIEW

File: cache/network/download.go [hunk 1]
```diff
@@ -45,31 +47,18 @@ func Download(params DownloadParams, logger log.Logger) (matchedKey string, err
...
+	downloadErr := downloadFile(retryableHTTPClient.StandardClient(), restoreResponse.URL, params.DownloadPath)
+	if downloadErr != nil {
+		return "", fmt.Errorf("failed to download archive: %w", downloadErr)
 	}
 
 	return restoreResponse.MatchedKey, nil
 }
+
+func downloadFile(client *http.Client, url string, dest string) error {
+	downloader := got.New()
```
Comment (zsolt-marta-bitrise) [inline cache/network/download.go:60]: Does this lib have any kind of debug logging feature? We have a debug log param in the steps and would be great to support that

Comment (duplabe) [reply cache/network/download.go:60]: As we discussed, we go with the current debug logs (we log with the underlying HTTP client)

**Review by zsolt-marta-bitrise — APPROVED** [review]

File: cache/network/download.go [hunk 2]
```diff
@@ -45,31 +47,18 @@ func Download(params DownloadParams, logger log.Logger) (matchedKey string, err
...
-	defer func(respBody io.ReadCloser) {
-		err := respBody.Close()
-		if err != nil {
-			logger.Warnf(err.Error())
-		}
-	}(respBody)
-	_, err = io.Copy(file, respBody)
-	if err != nil {
-		return "", fmt.Errorf("failed to save archive to disk: %w", err)
+	downloadErr := downloadFile(ctx, retryableHTTPClient.StandardClient(), restoreResponse.URL, params.DownloadPath)
```
Comment (rescuerunaways) [inline cache/network/download.go:51]: hmm.. in all other places in the api we use DefaultClient, i.e. this way:
retryableHTTPClient.Do(), which is hashicorp's cleanhttpclient with some default retries/backoffs

here we are passing StandardClient, which is Go's default client

is this intentional?

File: cache/network/download.go [hunk 3]
```diff
@@ -45,31 +47,18 @@ func Download(params DownloadParams, logger log.Logger) (matchedKey string, err
...
 	}
 
 	return restoreResponse.MatchedKey, nil
 }
+
+func downloadFile(ctx context.Context, client *http.Client, url string, dest string) error {
+	downloader := got.New()
+	downloader.Client = client
+
+	return downloader.Do(got.NewDownload(ctx, url, dest))
```
Comment (rescuerunaways) [inline cache/network/download.go:63]: 😂  this looks way easier than a fork

File: cache/network/download.go (hunk 2 above)
Comment (rescuerunaways) [reply cache/network/download.go:51]: i assume the lib expects standard lib Go's client

**Review by rescuerunaways — APPROVED** [review]

File: cache/network/download.go (hunk 2 above)
Comment (duplabe) [reply cache/network/download.go:51]: yes, it requires a standard library *http.Client. `StandardClient()` gives back a wrapped *http.Client that is using the retry logic. So, the downloader is using a retry client under the hood to download each range. We tested that if one of the ranges download fails the client will retry it.

**Pull request merged**